### ISSUE TITLE
Zookeeper "zxid"  error when restarting zookeeper container on a different swarm machine 

### ISSUE THREAD

**Issue by frranck** [wurstmeister/kafka-docker#374]
Hello,
If I'm using your [docker-compose-swarm.yml 
](https://github.com/wurstmeister/kafka-docker/blob/master/docker-compose-swarm.yml)
with 2 docker hosts in the swarm (1 master 1 worker) and scale the kafka to 2.
when zookeeper is moving from one host to the other. (i.e by restarting the container)

I'm getting these error in zookeeper:

```
2018-08-01 10:24:13,735 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxnFactory@192] - Accepted socket connection from /10.0.5.3:53856
2018-08-01 10:24:13,736 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:ZooKeeperServer@901] - Refusing session request for client /10.0.5.3:53856 as it has seen zxid 0x7a our last zxid is 0x0 client must try another server
2018-08-01 10:24:13,737 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxn@1008] - Closed socket connection for client /10.0.5.3:53856 (no session established for client)


```

and kafka stops working. (until I restart kafka as mentionned here : https://github.com/wurstmeister/kafka-docker/issues/315 )

Is something missing in this configuration ?

**Comment by sscaling** [comment]
I can't really vouch for the completeness of the referenced docker-compose file as it was provided by the community. It's really just meant as a place to get started, and not recommended for production use due to lack of persistence and no HA zookeeper cluster etc.

If you're just using the docker compose file as-is, then it's possibly something along the lines of the kafka zookeeper client trying to reconnect to zookeeper using the existing connection, but zookeeper has started up in a new location with no previous state.

Locally (without Swarm) I can reproduce by just killing and restarting zookeeper

```
kafka_1      | [2018-08-01 12:49:59,185] INFO Opening socket connection to server zookeeper/172.20.0.2:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn)
kafka_1      | [2018-08-01 12:49:59,186] INFO Socket connection established to zookeeper/172.20.0.2:2181, initiating session (org.apache.zookeeper.ClientCnxn)
zookeeper_1  | 2018-08-01 12:49:59,187 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxnFactory@192] - Accepted socket connection from /172.20.0.3:60142
kafka_1      | [2018-08-01 12:49:59,188] INFO Unable to read additional data from server sessionid 0x164f58763650000, likely server has closed socket, closing socket connection and attempting reconnect (org.apache.zookeeper.ClientCnxn)
zookeeper_1  | 2018-08-01 12:49:59,187 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:ZooKeeperServer@901] - Refusing session request for client /172.20.0.3:60142 as it has seen zxid 0x1f our last zxid is 0x0 client must try another server
zookeeper_1  | 2018-08-01 12:49:59,187 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxn@1008] - Closed socket connection for client /172.20.0.3:60142 (no session established for client)
kafka_1      | [2018-08-01 12:50:00,651] INFO Opening socket connection to server zookeeper/172.20.0.2:2181. Will not attempt to authenticate using SASL (unknown error) (org.apache.zookeeper.ClientCnxn)
```

A quick google finds similar issues, such as https://stackoverflow.com/questions/45804955/zookeeper-refuses-kafka-connection-from-an-old-client

So to solve your problem, I'd recommend using zookeeper with persistence and preferably in replicated mode (so quorum can withstand a certain amount of failures).

**Comment by frranck** [comment]
Thanks for your feedback.
Wouldn't it make sense to include a more elaborate version of this docker-compose file to the project ?

**Comment by sscaling** [comment]
It's a starting point - partly because Kafka depends on zookeeper (which is external to this project) and partly because the execution environment is user specific.

If you can think-of/find anything generic relating to the Kafka side we'd definitely be grateful for a PR.

I think this issue can be closed now?

**Comment by nastarangithub** [comment]
> Hello,
> If I'm using your [docker-compose-swarm.yml ](https://github.com/wurstmeister/kafka-docker/blob/master/docker-compose-swarm.yml)
> with 2 docker hosts in the swarm (1 master 1 worker) and scale the kafka to 2.
> when zookeeper is moving from one host to the other. (i.e by restarting the container)
> 
> I'm getting these error in zookeeper:
> 
> ```
> 2018-08-01 10:24:13,735 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxnFactory@192] - Accepted socket connection from /10.0.5.3:53856
> 2018-08-01 10:24:13,736 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:ZooKeeperServer@901] - Refusing session request for client /10.0.5.3:53856 as it has seen zxid 0x7a our last zxid is 0x0 client must try another server
> 2018-08-01 10:24:13,737 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxn@1008] - Closed socket connection for client /10.0.5.3:53856 (no session established for client)
> ```
> 
> and kafka stops working. (until I restart kafka as mentionned here : #315 )
> 
> Is something missing in this configuration ?



> Hello,
> If I'm using your [docker-compose-swarm.yml ](https://github.com/wurstmeister/kafka-docker/blob/master/docker-compose-swarm.yml)
> with 2 docker hosts in the swarm (1 master 1 worker) and scale the kafka to 2.
> when zookeeper is moving from one host to the other. (i.e by restarting the container)
> 
> I'm getting these error in zookeeper:
> 
> ```
> 2018-08-01 10:24:13,735 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxnFactory@192] - Accepted socket connection from /10.0.5.3:53856
> 2018-08-01 10:24:13,736 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:ZooKeeperServer@901] - Refusing session request for client /10.0.5.3:53856 as it has seen zxid 0x7a our last zxid is 0x0 client must try another server
> 2018-08-01 10:24:13,737 [myid:] - INFO  [NIOServerCxn.Factory:0.0.0.0/0.0.0.0:2181:NIOServerCnxn@1008] - Closed socket connection for client /10.0.5.3:53856 (no session established for client)
> ```
> 
> and kafka stops working. (until I restart kafka as mentionned here : #315 )
> 
> Is something missing in this configuration ?

You should persist kafka and zookeeper logs outside of docker. In order to do this, define a volume for both kafka and zookeeper services to mount a directory of the host machine to the directory containing the logs inside the service scope. This is a sample configuration:

zookeeper:
    image: wurstmeister/zookeeper:latest
    ports:
      - "2181:2181"
    hostname: zookeeper
    container_name: zookeeper
    volumes:
      - /docker/zookeeper:/opt/zookeeper-3.4.13/data
    networks:
      - net

kafka:
    hostname: kafka
    image: wurstmeister/kafka:latest
    depends_on:
      - zookeeper
    ports:
      - "9092:9092"
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /docker/kafka:/kafka
    networks:
      - net
